### ISSUE TITLE
inter-gpu communication

### ISSUE THREAD

**Issue by ShijianXu** [Spijkervet/SimCLR#13]
Hi, I have a question about the multi-gpu training.

The original SimCLR paper uses a large batch size and the good performance is greatly contributed to the large batch size. This is critical when computing the loss. However, in your implementation, it seems you didn't consider other sub-batches in the other gpus. Specifically, if you set the batch size as 1024 and use 8 gpus, then in each gpu, the sub-batch size is 128. According to your implementation, it seems that when computing the nt_xent loss, the number of samples used is just 128, not 1024. I think the correct implementation should use `all_gather` operations to get all the 1024 samples from all the gpus.

I am not sure about the actual implementation. If I understand something wrong, please correct me.

**Comment by Spijkervet** [comment]
Hi Shijian Xu,

Thank you for opening the issue. It is correct that the sim. matrix constructed in the loss function is done on the sub-batches on each GPU, and the loss is thus computed for every sub-batch and then aggregated in the way DataParallel/DistributedDataParallel aggregates losses. After taking a good look at the original implementation, this is indeed different: the original author first aggregates the data from every TPU ("replicas"), before calculating the loss.

Feel free to submit a PR if you have the right solution for this.

**Comment by Spijkervet** [comment]
Hi Shijian,

I've looked into the matter more closely. For DistributedDataParallel training, you are right about gathering tensors from all GPU's. I've changed it in [this commit](https://github.com/Spijkervet/SimCLR/commit/f75dc28ce1cf707124e77e99e5a7e4b3e967c734 ) so samples from all GPU's are gathered in `./modules/NT_Xent.py`

With DataParallel training, everything seems to go fine, since only the model is wrapped in the parallel module. For convenience, I added a flag in the config for DP training.
